File: node/src/aktcache.test.ts
import test from "ava";
import { newGroup, Getter, getGroup } from "./aktcache";

const db = {
  akt: "123",
  akita: "321",
  akitasummer: "213",
};

test("test AktCache get", async (t) => {
  const loadCounts: { [propsName: string]: number } = {};

  const getter: Getter = {
    Get(key: string) {
      const v = db[key];
      if (v) {
        const counts = loadCounts[key];
        if (!counts) {
          loadCounts[key] = 0;
        }
        loadCounts[key] += 1;
        return v;
      }

      return null;
    },
  };

  const akt = await newGroup("num", getter, 2048);

  const keys = Object.keys(db);

  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    const v = await akt.get(key);
    t.is(v, db[key]);
    await akt.get(key);
    t.false(loadCounts[key] > 1);
  }

  const v = await akt.get("test");

  t.is(v, null);

  t.pass();
});

test("test AktCache getGroup", async (t) => {
  const groupName = "num1";

  const getter: Getter = {
    Get(key: string) {
      return null;
    },
  };

  const akt = await newGroup(groupName, getter, 2048);

  const group = await getGroup(groupName);

  t.deepEqual(group, akt);

  const notGroup = await getGroup("1234");

  t.is(notGroup, null);

  t.pass();
});

File: go/aktcache/aktcache_test.go
package aktcache

import (
	"fmt"
	"reflect"
	"testing"
)

var db = map[string]string{
	"akt":         "123",
	"akita":       "321",
	"akitasummer": "213",
}

func TestGetter(t *testing.T) {
	var f Getter = GetterFunc(func(key string) ([]byte, error) {
		return []byte(key), nil
	})

	expect := []byte("key")
	if v, _ := f.Get("key"); !reflect.DeepEqual(v, expect) {
		t.Fatalf("callback failed")
	}
}

func TestGet(t *testing.T) {
	loadCounts := make(map[string]int, len(db)) // 统计get调用次数
	akt := NewGroup("num", 2048, GetterFunc(
		func(key string) ([]byte, error) {
			if v, ok := db[key]; ok {
				if _, ok := loadCounts[key]; !ok {
					loadCounts[key] = 0
				}
				loadCounts[key] += 1
				return []byte(v), nil
			}
			return nil, fmt.Errorf("%s not exist", key)
		},
	))

	for k, v := range db {
		// 如果cache值不正确
		if view, err := akt.Get(k); err != nil || view.String() != v {
			t.Fatalf("failed to get value")
		}

		// 如果get调用超过一次
		if _, err := akt.Get(k); err != nil || loadCounts[k] > 1 {
			t.Fatalf("cache %s miss", k)
		}
	}

	if view, err := akt.Get("test"); err == nil {
		t.Fatalf("this value is should be empty, but got %s", view)
	}
}

func TestGetGroup(t *testing.T) {
	groupName := "num"
	NewGroup(groupName, 2048, GetterFunc(
		func(key string) (bytes []byte, err error) { return }))
	if group := GetGroup(groupName); group == nil || group.name != groupName {
		t.Fatalf("group %s not exist", groupName)
	}

	if group := GetGroup(groupName + "111"); group != nil {
		t.Fatalf("expect nil, but %s got", group.name)
	}
}
